### ISSUE TITLE
Test failures in 4.0

### ISSUE THREAD

**Issue by yrro** [wincent/command-t#251]
Several tests fail with one of two errors. I'm reporting them together because this smells like a problem with my environment rather than command-t itself, but I'm not a ruby person and haven't had any luck figuring this out myself.

The errors are:

```
  1) CommandT::Controller accept selection opens relative paths inside the working directory
     Failure/Error: stub(finder).path = anything

     NoMethodError:
       undefined method `stub' for #<RSpec::ExampleGroups::CommandTController::AcceptSelection:0x000000027d02b0>
     # ./spec/command-t/controller_spec.rb:62:in `stub_finder'
     # ./spec/command-t/controller_spec.rb:12:in `block (3 levels) in <top (required)>'
```

and

```
  5) CommandT::Finder::BufferFinder sorted_matches_for method returns an empty array when no matches
     Failure/Error: any_instance_of(CommandT::Scanner::BufferScanner, :paths => @paths)

     NoMethodError:
       undefined method `any_instance_of' for #<RSpec::ExampleGroups::CommandTFinderBufferFinder::SortedMatchesForMethod:0x000000027a2ba8>
       Did you mean?  an_instance_of
     # ./spec/command-t/finder/buffer_finder_spec.rb:9:in `block (2 levels) in <top (required)>'
```

I'm building with Ruby 2.3.1p112, RR 1.2.0 and RSpec 3.5.0c3e0m0s0-1. Maybe these are just too old--they're the versions currently in Debian unstable.

**Comment by wincent** [comment]
The test suite is unfortunately a bit of a tire fire at this point. It's value is pretty low (lots of mocking, very fragile, breaks often, has seldom [never?] caught any bugs) and there have been long periods of time where breakage has gone undetected. It's also pretty sensitive to changes in its dependencies, I suspect.

It really needs a giant refactor to make it more testable, but that is non-trivial given how the current architecture is so intimately tied to the Vim runtime. See #220 for a long-running project to rearchitect Command-T in terms of a separate background process plus a thin VimL layer on the Vim side. As part of this, it should become a lot easier to test (additionally, it should become a lot easier to install because it will no longer be sensitive to Ruby versions in the way the current version is, and it should also make it easier to implement scanning behavior that doesn't block the UI, especially when running with Vim 8 where we have access to async communication features).

TL;DR: I periodically fix test breakages, but the test suite is likely going to remain a tire fire for a while.

**Comment by wincent** [comment]
Suite is green in 5.0. Going to close this for now.
